File: app/src/main/java/it/liceoarzignano/bold/utils/ContentUtils.java
package it.liceoarzignano.bold.utils;

import android.content.Context;
import android.content.res.Resources;

import com.firebase.jobdispatcher.FirebaseJobDispatcher;
import com.firebase.jobdispatcher.GooglePlayDriver;
import com.firebase.jobdispatcher.Job;
import com.firebase.jobdispatcher.Lifetime;
import com.firebase.jobdispatcher.RetryStrategy;
import com.firebase.jobdispatcher.Trigger;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import it.liceoarzignano.bold.R;
import it.liceoarzignano.bold.events.Event2;
import it.liceoarzignano.bold.events.EventsHandler;
import it.liceoarzignano.bold.events.EventsJobUtils;
import it.liceoarzignano.bold.marks.Mark2;
import it.liceoarzignano.bold.marks.MarksHandler;


public final class ContentUtils {

    private ContentUtils() {
    }

    /**
     * Get array of subjects with at least one mark for averages list
     *
     * @return array of subjects
     */
    public static String[] getAverageElements(Context context, int filter) {
        int size = 0;
        MarksHandler handler = MarksHandler.getInstance(context);

        List<Mark2> list = handler.getFilteredMarks(null, filter);
        ArrayList<String> elements = new ArrayList<>();

        for (Mark2 mark : list) {
            if (!elements.contains(mark.getSubject())) {
                elements.add(mark.getSubject());
                size++;
            }
        }

        return elements.toArray(new String[size]);
    }

    /**
     * Get event category description from int
     *
     * @param category: event icon value
     * @return category name
     */
    public static String eventCategoryToString(Context context, int category) {
        switch (category) {
            case 0:
                return context.getString(R.string.events_test);
            case 1:
                return context.getString(R.string.event_school);
            case 2:
                return context.getString(R.string.event_birthday);
            case 3:
                return context.getString(R.string.event_homework);
            case 4:
                return context.getString(R.string.event_reminder);
            case 5:
                return context.getString(R.string.event_meeting);
            default:
                return context.getString(R.string.event_other);
        }
    }

    /**
     * Fetch all the upcoming events and create a description
     *
     * @return content for events notification
     */
    public static String getTomorrowInfo(Context context) {
        Resources res = context.getResources();
        String content = "";

        int categories[] = new int[] {
                0 /* test */, 0 /* atSchool */, 0 /* bday */,
                0 /* homeworks */, 0 /* reminder */, 0 /* meeting */,
                0 /* others */
        };

        int messages[] = new int[] {
                R.plurals.notification_test, R.plurals.notification_school,
                R.plurals.notification_birthday, R.plurals.notification_homework,
                R.plurals.notification_reminder, R.plurals.notification_meeting,
                R.plurals.notification_other
        };

        EventsHandler handler = EventsHandler.getInstance(context);
        List<Event2> events = handler.getTomorrow();

        if (events.isEmpty()) {
            return null;
        }

        // Get data
        for (Event2 event : events) {
            categories[event.getCategory()]++;
        }

        // Build message
        for (int counter = 0; counter < categories.length; counter++) {
            if (categories[counter] > 0) {
                content = eventInfoBuilder(res, content, categories[counter], messages[counter]);
            }
        }
        content += " " + res.getString(R.string.notification_message_end);

        return content;
    }

    /**
     * Build part of summary for notification (only the given category)
     *
     * @param res to fetch strings
     * @param orig original message
     * @param size quantity of events
     * @param id string id
     * @return updated message
     */
    private static String eventInfoBuilder(Resources res, String orig, int size, int id) {
        String val = orig.isEmpty() ?
                res.getQuantityString(R.plurals.notification_message_first, size, size) :
                orig + res.getQuantityString(R.plurals.notification_message_half, size, size);

        return val + " " + res.getQuantityString(id, size, size);
    }

    public static void makeEventNotification(Context context) {
        Calendar calendar = Calendar.getInstance();
        int hr = calendar.get(Calendar.HOUR_OF_DAY);
        int userPreference = 21;

        switch (PrefsUtils.getEventsNotificationTime(context)) {
            case "0":
                userPreference = 6;
                break;
            case "1":
                userPreference = 15;
                break;
        }

        int when = (hr >= userPreference ? hr - userPreference + 1 : 24 + userPreference - hr)
                * 60 * 60;

        FirebaseJobDispatcher dispatcher = new FirebaseJobDispatcher(new GooglePlayDriver(context));
        Job job = dispatcher.newJobBuilder()
                .setService(EventsJobUtils.class)
                .setTag("Bold_EventsJob")
                .setLifetime(Lifetime.FOREVER)
                .setTrigger(Trigger.executionWindow(when, when + 120))
                .setRetryStrategy(RetryStrategy.DEFAULT_EXPONENTIAL)
                .setReplaceCurrent(false)
                .build();

        dispatcher.mustSchedule(job);
    }
}

File: app/src/main/kotlin/it/liceoarzignano/bold/utils/ContentUtils.kt
package it.liceoarzignano.bold.utils

import android.content.Context
import android.content.res.Resources
import com.firebase.jobdispatcher.*
import it.liceoarzignano.bold.R
import it.liceoarzignano.bold.events.EventsHandler
import it.liceoarzignano.bold.events.EventsJobUtils
import it.liceoarzignano.bold.marks.MarksHandler
import it.liceoarzignano.bold.settings.AppPrefs
import java.util.*

object ContentUtils {

    fun getAverageElements(context: Context, filter: Int): Array<String> {
        val handler = MarksHandler.getInstance(context)

        val list = handler.getFilteredMarks("", filter)
        val elements = ArrayList<String>()

        for (item in list) {
            if (!elements.contains(item.subject)) {
                elements.add(item.subject)
            }
        }

        return elements.toTypedArray()
    }

    fun eventCategoryToString(context: Context, category: Int): String =
            when (category) {
                0 -> context.getString(R.string.events_test)
                1 -> context.getString(R.string.event_school)
                2 -> context.getString(R.string.event_birthday)
                3 -> context.getString(R.string.event_homework)
                4 -> context.getString(R.string.event_reminder)
                5 -> context.getString(R.string.event_meeting)
                else -> context.getString(R.string.event_other)
    }

    fun getTomorrowInfo(context: Context): String {
        val res = context.resources
        var content = ""

        val categories = intArrayOf(0 /* test */, 0 /* atSchool */,
                0 /* bday */, 0 /* homeworks */,
                0 /* reminder */, 0 /* meeting */,
                0 /* others */)

        val messages = intArrayOf(R.plurals.notification_test, R.plurals.notification_school,
                R.plurals.notification_birthday, R.plurals.notification_homework,
                R.plurals.notification_reminder, R.plurals.notification_meeting,
                R.plurals.notification_other)

        val handler = EventsHandler.getInstance(context)
        val events = handler.getDayEvents(1)

        if (events.isEmpty()) {
            return ""
        }

        // Get data
        for (event in events) {
            categories[event.category]++
        }

        // Build message
        categories.indices
                .asSequence()
                .filter { categories[it] > 0 }
                .forEach { content = eventInfoBuilder(res, content, categories[it], messages[it]) }

        content += " " + res.getString(R.string.notification_message_end)

        return content
    }

    private fun eventInfoBuilder(res: Resources, orig: String, size: Int, id: Int): String {
        val message =
                if (orig.isEmpty())
                    res.getQuantityString(R.plurals.notification_message_first, size, size)
                else
                    orig + res.getQuantityString(R.plurals.notification_message_half, size, size)

        return message + " " + res.getQuantityString(id, size, size)
    }

    fun makeEventNotification(context: Context) {
        val calendar = Calendar.getInstance()
        val hr = calendar.get(Calendar.HOUR_OF_DAY)
        var userPreference = 21
        val prefs = AppPrefs(context)
        if (!prefs.get(AppPrefs.KEY_NOTIF_EVENT, true)) {
            return
        }

        when (prefs.get(AppPrefs.KEY_NOTIF_EVENT_TIME, "2")) {
            "0" -> userPreference = 6
            "1" -> userPreference = 15
        }

        val time =
                (if (hr >= userPreference)
                    hr - userPreference + 1
                else
                    24 + userPreference - hr) * 60 * 60

        val dispatcher = FirebaseJobDispatcher(GooglePlayDriver(context))
        val job = dispatcher.newJobBuilder()
                .setService(EventsJobUtils::class.java)
                .setTag("Bold_EventsJob")
                .setLifetime(Lifetime.FOREVER)
                .setTrigger(Trigger.executionWindow(time, time + 120))
                .setRetryStrategy(RetryStrategy.DEFAULT_EXPONENTIAL)
                .setReplaceCurrent(false)
                .build()

        dispatcher.mustSchedule(job)
    }

    fun getHeaderTitle(res: Resources, time: Time): Pair<String, String> {
        val diff = time.diff(Time())

        val title = when {
            // Past weeks
            diff < -7 -> res.getQuantityString(R.plurals.events_time_past, 2, diff / -7)
            // Last week
            diff in -7..-2 -> res.getQuantityString(R.plurals.events_time_past,
                    1, time.getWeekDay()).upperCaseFirstChar()
            // Yesterday
            diff == -1 -> res.getString(R.string.events_time_yesterday)
            // Today
            diff == 0 -> res.getString(R.string.events_time_today)
            // Tomorrow
            diff == 1 -> res.getString(R.string.events_time_tomorrow)
            // This week
            diff in 2..7 -> time.getWeekDay().upperCaseFirstChar()
            // Future weeks
            else -> res.getString(R.string.events_time_future, diff / 7)
        }

        val description = time.format(res.getString(R.string.events_time_default_format))
                .upperCaseFirstChar()

        return Pair(title, description)
    }


    private fun String.upperCaseFirstChar() =
            "${substring(0, 1).toUpperCase()}${substring(1, length)}"
}
